_posts/2024-02-14-securing-wildfly-apps-okta-openshift.adoc

```diff
@@ -0,0 +1,267 @@
+
+. Examine the https://github.com/PrarthonaPaul/elytron-examples/blob/simple-webapp-okta/simple-webapp-okta/src/main/webapp/WEB-INF/oidc.json[oidc.json] file. The oidc.json is used to configure the native OIDC client subsystem.
+```
+{
+    "client-id" : "${env.CLIENT_ID}",
+    "provider-url" : "${env.OIDC_PROVIDER_URL}",
+    "public-client" : "false",
+    "principal-attribute" : "IDTClaim",
+    "ssl-required" : "EXTERNAL",
+    "credentials" : {
+        "secret" : "${env.CLIENT_SECRET}"
+    }
+}
+```
+* Client_ID is the unique identifier for our client used by the Okta OpenID provider. Usually for Keycloak, you can create your own client, but for Okta, you are assigned an id. As a result, we are using environment variables to specify this. We will set the value in 
+* The provider URL, which is the URL for the authorization server that we created, is specified as an environment variable. We will set its value in the helm configuration.
+* When public-client set to false, client credentials are sent when communicating with the OpenID provider.
+* We specify that the user name of the identity, which in our case is alice, is to be used as the principal for the identity. We are extracting this information here using a custom claim in the ID token.
+* When ssl-required is set to EXTERNAL, only the communication with external clients happens over HTTPs
+
+== Get the Application URL
+
+Once the WildFly server has been provisioned, use the following command to find the URL for your example application:
+
+```
+SIMPLE_WEBAPP_OKTA_URL=https://$(oc get route oidc-app --template='{{ .spec.host }}') &&
+echo "" &&
+echo "Application URL: $SIMPLE_WEBAPP_OKTA_URL/simple-webapp-okta"  &&
+echo "Valid redirect URI: $SIMPLE_WEBAPP_OKTA_URL/simple-webapp-okta/secured/" &&
+echo ""
+```
+We’ll make use of these URLs in the next two sections.
+
+== Finish Configuring Okta
+
+Go back to the _General Settings_ for your application and click on _Edit_.  add the `Valid redirect URI` under the `Sign-in redirect URIs` and check the `Allow wildcard * in login URI redirect` field and hit Save.
+
+== Access the App
+Now we can access our application using the *Application Url* from the previous section. 
+Click on "Access Secured Servlet". 
+
+Now you will be redirected to the login page for Okta. Login using Alice. You will be prompted for the username. Although we set the username to be *alice@example.org*, we can just input alice here. You will be presented with three options for loggin in. Choose `Password` and enter the password you selected for Alice. 
+
+Once you have entered the correct password, you will be prompted to perform 2 factor authentication using the Okta Verify mobile application. You can go ahead and follow the instructions to set it up and once you complete the 2 factor authentication, you will be redirected to the secured page. 
+
+// == Disabling 2 Factor Authentication
```

Comment:
Commenting it out for now because I think Okta changes their sign in policy to not make 2FA the default policy.